proposals/client-interceptors.md

```diff
@@ -0,0 +1,312 @@
+```
+
+### Interceptor API
+
+An interceptor is a function which takes an `options` object and a `nextCall` function and returns an `InterceptingCall` object:
+
+```javascript
+/**
+ * @param {Object} options
+ * @param {string} options.hostname The server host to call
+ * @param {Object} options.credentials Credentials used to authenticate with the server
+ * @param {MethodDescriptor} options.methodDiscriptor A container of properties describing the call (see below)
+ * @param {Function} nexCall Constructs the next interceptor in the chain
+ * @returns {InterceptingCall} 
+ */
+var interceptor = function(options, nextCall) {
+  return new InterceptingCall(nexCall(options));
+}
+```
+
+An interceptor function allows developers to modify the call options, store any call-scoped values needed, and define interception methods.
+
+The interceptor function must return an `InterceptingCall` object. An `InterceptingCall` represents an element in the interceptor chain. Any intercepted methods are implemented in the requester object, an optional parameter to the constructor. Returning `new InterceptingCall(nextCall(options))` will satisfy the contract (but provide no interceptor functionality).
+
+Most interceptors will define a `requester` object (described below) which is passed to the the `InterceptingCall` constructor: `return new InterceptingCall(nextCall(options), requester);`
+
+The `options` argument to the `nextCall` function includes options accepted by `rpcCall` and `serverStreaming` constructor. Modifying the options that are passed to the `nextCall` function will have the effect of changing the options passed to the underlying `rpcCall`/`serverStreaming` constructor.
+
+Additionally, the `options` argument includes a `MethodDescriptor` object. The `MethodDescriptor` is a container for properties of the call which are used internally and may also be useful to interceptors:
+
+```javascript
+/**
+ * @param {string} name The name of the call, i.e. 'myCall' 
+ * @param {string} serviceName The name of the service, i.e. 'MyService'
+ * @param {string} path The full path of the call, i.e. '/MyService/MyCall'
+ * @param {MethodType} methodType One of two types: MethodType.UNARY or MethodType.SERVER_STREAMING
+ * @param {grpc.web.AbstractClientBase.MethodInfo} methodInfo Properties containing the serialize and deserialize functions
+ */
+MethodDescriptor(name, serviceName, path, methodType, methodInfo)
+```
+*Do not modify the `options.methodDescriptor` object. The MethodDescriptor passed to the interceptor options is not consumed by the underlying gRPC code and changes will only affect downstream interceptors*
+
+A `requester` object is a plain Javascript object implementing zero or more outbound interception methods:
+
+```javascript
+/**
+ * An interception method called before an outbound call has started.
+ * @param {grpc.web.Metadata} metadata The call's outbound metadata (request headers).
```

Comment:
Is grpc.web.Metadata a new class?

Reply:
No, I stole this from the TypeScript definition:
https://github.com/grpc/grpc-web/blob/master/packages/grpc-web/index.d.ts#L3
I probably should have left it as `Object<string, string>`